### PR TITLE
Feature/exclude patterns

### PR DESCRIPTION
**Pull request by khalwat**
Adds the ability to have an optional `exclusions-patterns.txt` file in the project root, which are regular expressions that are matched against the incoming message.

If there's a match, it does not post the message to Twitter.

### CODE REVIEW

File: Dockerfile.armhf [hunk 1]
```diff
@@ -33,7 +33,9 @@ RUN \
     echo "${TIME_ZONE}" > /etc/timezone && date && \
     ln -s /usr/bin/dumb-init /sbin/dinit && \
     rm -rf /var/cache/apk/*
-    
+
+COPY ./exclusion-patterns.txt /srv/exclusion-patterns.txt
```
[the remark this answers is not in the document]
Comment (khalwat) [reply Dockerfile.armhf]: Fixed the one, thanks for the suggestion, this is a much better way to do it!

File: app/publisher/publisher.go [hunk 2]
```diff
@@ -18,18 +19,39 @@ type Interface interface {
 }
 
 // Stdout implements publisher.Interface and sends to stdout
-type Stdout struct{}
+type Stdout struct{
+	ExcludeList []string
+}
+
+// CheckExclusionList checks the exclusion list for matches
+func CheckExclusionList(excludes []string, msg string) bool {
```
[the remark this answers is not in the document]
Comment (khalwat) [reply app/publisher/publisher.go:27]: Added TestExclusionPatterns test

File: app/main.go [hunk 3]
```diff
@@ -74,16 +75,24 @@ func main() {
 }
 
 func setup(o opts) (n notifier, p publisher.Interface, err error) {
+	content, err := ioutil.ReadFile("exclusion-patterns.txt")
+	if err != nil {
+		log.Print("[WARN] could not read 'exclusion-patterns.txt' file")
+	}
+	lines := strings.Split(string(content), "\n")
```
Comment (umputun) [inline app/main.go:82]: this is unsafe. in case of err we don't want to touch the content.

File: app/main_test.go [hunk 4]
```diff
@@ -126,6 +127,24 @@ func TestFormat(t *testing.T) {
 	}
 }
 
+func TestExclusionPatterns(t *testing.T) {
+	tbl := []struct {
+		msg  		string
+		result  	bool
+		ExcludeList []string
+	}{
+		{"All I hear is blah blah blah", true, []string{"blah"}},
```
Comment (umputun) [inline app/main_test.go:136]: the match is regex, but tests don't have anything using regex match

File: app/publisher/publisher.go [hunk 5]
```diff
@@ -18,18 +19,39 @@ type Interface interface {
...
+	for _, value := range excludes {
+		if len(value) > 0 && !strings.HasPrefix(value, "#") {
+			if match, _ := regexp.MatchString(strings.ToLower(value), strings.ToLower(msg)); match {
```
Comment (umputun) [inline app/publisher/publisher.go:30]: ignoring errors generally makes me worry. In this case, I see no reason to suppress it. In rare cases, it makes sense but documenting the reason (in a comment)  will help to follow the logic.

File: exclusion-patterns.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+# Put any regular expressions to exclude messages here, one regex per line
```
Comment (umputun) [inline exclusion-patterns.txt:1]: will be nice to have a few commented out examples of trivial and less trivial cases

**Comment by khalwat** [comment]
> example of mapping in docker-compose.yml (commented out)

Should this be commented out though? If the file always exists (with commented examples in it) then it should be okay to just leave the bind mapping in the `docker-compose.yaml` file?

**Comment by umputun** [comment]
up to you. I don't have a problem with uncommented volume mapping and commented rules file

File: app/main.go (hunk 3 above)
Comment (khalwat) [reply app/main.go:83]: Ok, I'm initializing content to an empty slice of bytes now

File: app/publisher/publisher.go (hunk 5 above)
Comment (khalwat) [reply app/publisher/publisher.go]: Okay it's logging errors from regexp.MatchString now

File: app/main_test.go (hunk 4 above)
Comment (khalwat) [reply app/main_test.go]: Added regex's to the tests

File: exclusion-patterns.txt (hunk 6 above)
Comment (khalwat) [reply exclusion-patterns.txt]: Added a few basic regex's to the examples, with a link to a more in-depth tutorial on regular expressions

File: app/main.go [hunk 7]
```diff
@@ -74,16 +75,25 @@ func main() {
 }
 
 func setup(o opts) (n notifier, p publisher.Interface, err error) {
+	content, err := ioutil.ReadFile("exclusion-patterns.txt")
+	if err != nil {
+		log.Print("[WARN] could not read 'exclusion-patterns.txt' file")
```
Comment (umputun) [inline app/main.go:80]: keeping context is important, but showing the actual error is even more important, `i.e. log.Printf("[WARN] could not read 'exclusion-patterns.txt' file, %v", err)`

File: app/publisher/publisher.go [hunk 8]
```diff
@@ -18,18 +19,43 @@ type Interface interface {
...
+}
+
+// CheckExclusionList checks the exclusion list for matches
+func CheckExclusionList(excludes []string, msg string) bool {
+	for _, value := range excludes {
+		if len(value) > 0 && !strings.HasPrefix(value, "#") {
+			match, err := regexp.MatchString(strings.ToLower(value), strings.ToLower(msg));
+			if err != nil {
+				log.Printf("[WARN] regexp.MatchString error: %s", err)
+			}
```
Comment (umputun) [inline app/publisher/publisher.go:33]: return false right away will make following the logic easier

File: app/main.go (hunk 7 above)
Comment (khalwat) [reply app/main.go:80]: Explicit sad path return added

File: app/publisher/publisher.go (hunk 8 above)
Comment (khalwat) [reply app/publisher/publisher.go:34]: Explicit sad path return added

File: app/main.go (hunk 7 above)
Comment (khalwat) [reply app/main.go]: Added logging of the error to the error message

**Comment by umputun** [comment]
LGTM

**Comment by khalwat** [comment]
Thanks for your patience with me on this!

**Pull request merged**